qiskit/circuit/quantumcircuitdata.py

```diff
@@ -30,17 +133,28 @@ def __getitem__(self, i):
         return self._circuit._data[i]
 
     def __setitem__(self, key, value):
-        instruction, qargs, cargs = value
-
-        if not isinstance(instruction, Instruction) and hasattr(instruction, "to_instruction"):
-            instruction = instruction.to_instruction()
-        if not isinstance(instruction, Instruction):
+        # For now (Terra 0.21), the `QuantumCircuit.data` setter is meant to perform validation, so
+        # we do the same qubit checks that `QuantumCircuit.append` would do.
```

Comment:
I assume the plan is to deprecate this in the future?

Reply:
Ideally, I suppose.  We'd need to work out exactly what API for low-level modification of `data` we want to support, and get that in place before we do anything like that.